File: 279-完全平方数.cpp
class Solution {
public:
    int numSquares(int n) {
        vector<int> nums = {0, 1};
        for(int i = 0; i < n - 1; i++){
            nums.push_back(INT_MAX);
        }
        
        for(int i = 2; i < n + 1; i++){
            for(int j = 1; j * j <= i; j++){
                nums[i] = min(nums[i], nums[i - j * j] + 1);
            }
        }

        return nums[n];
    }
};

File: 279-完全平方数.py
class Solution:
    def numSquares(self, n: int) -> int:
        nums = [0, 1] + [float('inf')] * (n - 1)   
        
        for i in range(2, n + 1):
            j = 1
            while j ** 2 <= i:
                nums[i] = min(nums[i], nums[i - j ** 2] + 1)
                j += 1
        
        return nums[n]
